### ISSUE TITLE
Game doesn't let some spells enter the stack despite having mana to cast them

### ISSUE THREAD

**Issue by Zerrisx** [magefree/mage#6698]
This is a thread for any spells that you are unable to be cast despite having the mana to pay for them.  Most of these will likely be spells with cost reductions where you can afford the reduced version but not the full version, or where your mana sources are somehow weird (moxen, Channel, spirit guides, creatures that generate variable amounts of mana, etc).

## Currently, the following have been reported:

- [ ] Hogaak, Arisen Necropolis - can't pay convoke #6680

- [ ] K'rrik, Son of Yawgmoth bug - can't use phyrexian ability unless you could have paid the full cost

- [ ] Cannot cast Price of Fame if Mana open is less than 4 #6685 

- [ ] Cannot cast spells if a Chrome Mox is untapped (reported in discord from multiple users)

**Comment by Hateful-Little-Bear** [comment]
Having this problem with Tasigur, the Golden Fang. I can't attempt to use delve to cast him from command zone.

![Screenshot (1)](https://user-images.githubusercontent.com/67252556/85247671-8f086f00-b41c-11ea-9dc3-ce5bd2efd191.png)

**Comment by LevelX2** [comment]
I would close all the duplicate tickets.

Or should we let them open until the related problem is resolved?
But because the needed changes are probably more general it's unneccessary work.

**Comment by JayDi85** [comment]
@LevelX2 no needs to close original issues -- it's contains additional info and users can find it by default search until real fix (github searches in active issues by default). Just add related links to that post or wait Zerrisx (he can collect and update that list too). When you fix card then close all related issues from that post too.

My logic: if I can find active issues with card name/ability then problem isn't fixed yet in xmage. Single tasks are much easier to find than shared lists. As example: feedback topics -- I need to open every founded feedback list to find is it fixed or not:
![shot_200622_191313](https://user-images.githubusercontent.com/8344157/85304177-810a2b00-b4bc-11ea-8ac0-c7a5d2e878b2.png)

**Comment by Zerrisx** [comment]
Should we also report cases of cards thinking they can be cast when they can't?  Some of them might have good reason, but they could also be indicative of issues.

For example, Willbender thinks it can be cast at all times - including during my opponent's turn with no mana sources available.
![image](https://user-images.githubusercontent.com/59861365/85312821-00780a00-b47d-11ea-8cd9-5c9d74e202d3.png)

**Comment by JayDi85** [comment]
Yes, report it too. Also you can report some potential bugged cards or combo that need to check. Just mark it as possibly bugged.

**Comment by serokellcao** [comment]
Wizard's Lightning doesn't work with Githu Lavarunner, but works with Burning Prophet.

Happened in one game, but not in another! Undefined behaviour is sad :(

**Comment by Zerrisx** [comment]
> 
> 
> Wizard's Lightning doesn't work with Githu Lavarunner, but works with Burning Prophet.
> 
> Happened in one game, but not in another! Undefined behaviour is sad :(

Weird - both cards worked with the discount for me; was able to cast Wizard's Lightning with just one mountain.

**Comment by Terminus42** [comment]
First off sorry for not having a screenshot showing the exact game state.

Trying to cast Expansion // Explosion's Expansion part with two floating blue mana and without untapped lands consistently produces an issue for me. The spell doesn't begin casting, quite possibly because the game calculates the required amount of floating mana incorrectly.

**Comment by LevelX2** [comment]
K'rrik, Son of Yawgmoth problem seems to be fixed, can't reproduce the problem.

**Comment by Zerrisx** [comment]
> 
> 
> K'rrik, Son of Yawgmoth problem seems to be fixed, can't reproduce the problem.
> 
> Same for Torgaar, Famine Incarnate .

K'rrik eems reproducible enough to me testing after the latest commit:

![image](https://user-images.githubusercontent.com/59861365/86489374-c71b8600-bd29-11ea-9032-f3d868ff7fd1.png)

Torgaar appears to be fixed.

**Comment by SpikesCafe-google** [comment]
I think this is probably related. [Kaleidostone](https://scryfall.com/card/con/137/kaleidostone) makes it think I can't cast any of the spells in my hand. If I tap the mana before casting the spell, it lets me cast them.
![Capture](https://user-images.githubusercontent.com/16656302/86492993-d4427000-bd3d-11ea-9066-268be3c1fd85.PNG)

**Comment by Zerrisx** [comment]
> 
> 
> First off sorry for not having a screenshot showing the exact game state.
> 
> Trying to cast Expansion // Explosion's Expansion part with two floating blue mana and without untapped lands consistently produces an issue for me. The spell doesn't begin casting, quite possibly because the game calculates the required amount of floating mana incorrectly.

Can't reproduce this one.

> 
> 
> I think this is probably related. [Kaleidostone](https://scryfall.com/card/con/137/kaleidostone) makes it think I can't cast any of the spells in my hand. If I tap the mana before casting the spell, it lets me cast them. If I have 5 lands untapped and could activate Kaleidostone, it also lets me cast them. On 1.4.44V0-beta17.

Yep, this one I can; adding to the list.

**Comment by LevelX2** [comment]
The problem does also exist for all spells with the "<Subtype> offering" keyword. 
Card can only be cast if **all** mana is available in the normal way.

**Comment by Zerrisx** [comment]
> 
> 
> The problem does also exist for all spells with the " offering" keyword.
> Card can only be cast if **all** mana is available in the normal way.

I swear I added that, but I think it got wiped out by somebody else's edit.  Adding it again.

**Comment by Zerrisx** [comment]
Major categories that seem to have issues right now: Storage Lands/artifacts, Mana Doublers, Permanents that can tap *other* permanents to add mana.

**Comment by LevelX2** [comment]
>     * Deathrite Shaman - unable to cast Faithless Looting off a Deathrite Shaman with a plains in my graveyard

That ability is no mana ability (because of the target). So it would mean to scan all abilities (instead of now only mana abilities) during calculation of available mana calculation. 
Don't know if we should do that.

**Comment by Zerrisx** [comment]
> 
> 
> > ```
> > * Deathrite Shaman - unable to cast Faithless Looting off a Deathrite Shaman with a plains in my graveyard
> > ```
> 
> That ability is no mana ability (because of the target). So it would mean to scan all abilities (instead of now only mana abilities) during calculation of available mana calculation.
> Don't know if we should do that.

Oh right, I always forget that one's *not* a mana ability.  Removing from the list.

**Comment by Zerrisx** [comment]
Finished going through all the mana abilities in my searches.

**Comment by acabaca24** [comment]
Yesterday my opponent couldn't activate his Rishadan Port in my upkeep when he was given priority. His untapped lands were Rishadan Port and Thespian Stage, I'm pretty sure this is related to this issue, that's why I didn't open a new one. Hopefully I'm right.

**Comment by Zerrisx** [comment]
> 
> 
> Yesterday my opponent couldn't activate his Rishadan Port in my upkeep when he was given priority. His untapped lands were Rishadan Port and Thespian Stage, I'm pretty sure this is related to this issue, that's why I didn't open a new one. Hopefully I'm right.

It could be, but I'd be a little surprised - most of these issues are due to bad cost-reduction or mana adding calculations, and Rishadan Port / Thespian's Stage don't have any of that.  I'd suggest writing it up separately.

**Comment by Zerrisx** [comment]
> 
> 
> First off sorry for not having a screenshot showing the exact game state.
> 
> Trying to cast Expansion // Explosion's Expansion part with two floating blue mana and without untapped lands consistently produces an issue for me. The spell doesn't begin casting, quite possibly because the game calculates the required amount of floating mana incorrectly.

Cannot reproduce.

**Comment by acabaca24** [comment]
> > Yesterday my opponent couldn't activate his Rishadan Port in my upkeep when he was given priority. His untapped lands were Rishadan Port and Thespian Stage, I'm pretty sure this is related to this issue, that's why I didn't open a new one. Hopefully I'm right.
> 
> It could be, but I'd be a little surprised - most of these issues are due to bad cost-reduction or mana adding calculations, and Rishadan Port / Thespian's Stage don't have any of that. I'd suggest writing it up separately.

Okay, I'll be doing that!

**Comment by Zerrisx** [comment]
I separated out the individual calculation card bugs from the combo bugs - most of the combos will be too complicated to reasonably calculate, so we should probably consider the cards involved as blank checks for the purposes of hard stops.  i.e. if you have Kydele, Chosen of Kruphix in play, you are considered to have infinite {C} mana.  Same for Treasonous Ogre and {R}, Skirge Familiar and {B}, Titan's Nest and {C} for colored non-X spells, and Pili-Pala for... I guess all colors of mana?  Pili-Pala calculations are a mess.

Short version is: tapped permanents, cards in any zone, and life are all common side effects of mana abilities and could have any number of additional replacement effects affecting them, so other mana abilities that are conditional on those traits produce indeterminate amounts of mana.

**Comment by LevelX2** [comment]
>     * Selvala, Explorer Returned - you should be able to cast spells that require up to G per player in the game from Selvala, but you can't cast a 1GG spell off a forest and Selvala in a two person game

Seems to work ok for me. You don't know if non land cards are revealed.

**Comment by Zerrisx** [comment]
> 
> 
> > ```
> > * Selvala, Explorer Returned - you should be able to cast spells that require up to G per player in the game from Selvala, but you can't cast a 1GG spell off a forest and Selvala in a two person game
> > ```
> 
> Seems to work ok for me. You don't know if non land cards are revealed.

So, Selvala, Explorer Returned is... *weird*.  Specifically, you can put a spell (say, Channel) on the stack and declare that you intend to pay for it by tapping Selvala and revealing two creatures, one from each player's library, as a mana ability.  You can do this even if you don't know that there's a creature on top of each deck.  You can do it even if you explicitly *do* know that one of the two decks doesn't have a creature on top.

However, if you do this and as a result can't pay for the spell, the entire cast is undone.  See this ruling on Selvala:

> 5/29/2014: _If you activate Selvala’s ability while casting a spell, and you discover you can’t produce enough mana to pay that spell’s costs, the spell is reversed. The spell returns to whatever zone you were casting it from. You may reverse other mana abilities you activated while casting the spell, but Selvala’s ability can’t be reversed. Whatever mana that ability produced will be in your mana pool and each player will have drawn a card._

When does it matter that you can do this with the spell on the stack instead of before casting it?

For one, because of Panglacial Wurm, everybody's favorite rule-breaking card.

Example: if you have Panglacial Wurm on top of your library and five forests plus an evolving wilds in play, you can crack a fetchland, start searching your library, put Panglacial Wurm on the stack, then cast it by revealing the card *below* Panglacial Wurm and the top of your opponent's library.  Since activating Selvala beforehand would result in drawing the Wurm, blind-activating like this is the only way to get the extra value off Panglacial Wurm, and MTG lets you do it.  If your opponent doesn't have the creature you needed, whoops, let's just pretend you never did that.

Currently, XMage does not allow this blind activation.
![image](https://user-images.githubusercontent.com/59861365/90527783-8b9d1580-e137-11ea-884e-499d61762902.png)
